### PR TITLE
First hand tracking

### PR DESCRIPTION
**Pull request by sangbin-sangbin**

### CODE REVIEW

File: test.py [hunk 1]
```diff
@@ -113,19 +128,23 @@ def normalize_points(points):
...
-start_recognizing_time_threshold = 3
-no_gesture_time = time.time()
-stop_recognizing_time_threshold = 3
+last_hand_time = time.time()
+
+wake_up_state = []
 
 while cap.isOpened():    
     time_threshold = cv2.getTrackbarPos('time','gesture recognition')/100
+    same_hand_threshold = cv2.getTrackbarPos('same_hand','gesture recognition')/1000
```
Comment (kmsgm) [inline test.py:145]: What's this for? I don't get it

File: test.py [hunk 2]
```diff
@@ -10,6 +10,36 @@
...
+    min_idx = -1
+    min_val = float('inf')
+    for i, [x, y] in enumerate(positions):
+        d = (x-prev_pos[0])**2 + (y-prev_pos[1])**2
+        if min_val > d:
+            min_idx = i
+            min_val = d
+
+    if min_val > same_hand_threshold:
+        return -1, prev_pos
```
Comment (kmsgm) [inline test.py:39]: Does `-1` mean `default` state?

File: test.py [hunk 3]
```diff
@@ -143,69 +162,126 @@ def normalize_points(points):
...
+                    recognizing_hand = landmark_lst
+                    lst, scale = normalize_points(landmark_lst)
+
+                    start = time.time_ns()//1000000
+                    res = list(model(torch.tensor([element for row in lst for element in row], dtype=torch.float)))
+                    end = time.time_ns()//1000000
+                    gesture_time += end - start
+                    gesture_num += 1
+
+                    p = max(res)
```
Comment (kmsgm) [inline test.py:191]: Please use more detailed variable name

**Review by kmsgm — COMMENTED** [review]
LGTM overall but please add more annotations for easy understanding.

File: test.py [hunk 4]
```diff
@@ -143,69 +162,126 @@ def normalize_points(points):
...
+                        elapsed_time = '0'
+                        prev_gesture = 'none'
+            else:
+                delete_list = []
+                wake_up_hands = []
+                for right_hand in right_hands:
+                    lst, scale = normalize_points(right_hand)
+
+                    res = list(model(torch.tensor([element for row in lst for element in row], dtype=torch.float)))
+                    p = max(res)
```
Comment (kmsgm) [inline test.py:236]: What's this for?

File: test.py [hunk 5]
```diff
@@ -10,6 +10,36 @@
...
+
+    positions = []
+    for landmark in hands:
+        center = get_center(landmark)
+        positions.append([ center[0], center[1] ])
+
+    min_idx = -1
+    min_val = float('inf')
+    for i, [x, y] in enumerate(positions):
+        d = (x-prev_pos[0])**2 + (y-prev_pos[1])**2
```
[the remark this answers is not in the document]
Comment (sangbin-sangbin) [reply test.py]: fixed

File: test.py (hunk 1 above)
Comment (sangbin-sangbin) [reply test.py:147]: fixed

File: test.py (hunk 2 above)
Comment (sangbin-sangbin) [reply test.py:40]: fixed

File: test.py (hunk 3 above)
Comment (sangbin-sangbin) [reply test.py]: fixed

File: test.py [hunk 6]
```diff
@@ -143,69 +162,126 @@ def normalize_points(points):
...
-                            subprocess.run('adb shell input tap 80 500', shell=True)
-                            print('left')
-                        elif gestures[state['gesture']] == 'select' and gestures[state['prev_gesture']] == 'default':
-                            subprocess.run('adb shell input tap 80 720', shell=True)
-                            print('select')
-                        elif gestures[state['gesture']] == 'exit' and gestures[state['prev_gesture']] == 'default':
-                            subprocess.run('adb shell input tap 80 820', shell=True)
-                            print('exit')
-                        state['prev_gesture'] = gesture_idx
+                if hand.classification[0].label == 'Left' or hand.classification[0].label == 'Right':
```
[the remark this answers is not in the document]
Comment (sangbin-sangbin) [reply test.py]: fixed

File: test.py (hunk 4 above)
Comment (sangbin-sangbin) [reply test.py]: fixed

File: test.py [hunk 7]
```diff
@@ -143,69 +165,136 @@ def normalize_points(points):
...
         landmark_num += 1
-        # Extract hand landmarks if available
+    
         if results.multi_hand_landmarks:
-            no_gesture_time = time.time()
-            # Get the coordinates of the index fingertip (landmark index 8)
-            hand_idx = -1
+            right_hands = []
             for idx, hand in enumerate(results.multi_handedness):
+                # label 'Left' means it is right hand because of 좌우반전
```
Comment (kmsgm) [inline test.py:172]: -> Left/Right Inversion

File: test.py [hunk 8]
```diff
@@ -43,26 +57,31 @@ def same_hand_tracking(hands, prev_pos, same_hand_threshold):
 
 def play_wav_file(file_name):
     pygame.mixer.init()
-    try:
-        pygame.mixer.music.load('./sound/'+file_name+'.wav')
-        pygame.mixer.music.play()
-        while pygame.mixer.music.get_busy():
-            pygame.time.Clock().tick(10)
-    except pygame.error as e:
-        print(f"Error playing WAV file: {e}")
```
Comment (kmsgm) [inline test.py:21]: Why did you delete try and except?

File: test.py [hunk 9]
```diff
@@ -299,6 +300,9 @@ def normalize_points(points):
     # Print Gesture    
     cv2.putText(frame, text_a, (frame.shape[1] // 2 + 230, frame.shape[0] // 2 - 220), cv2.FONT_HERSHEY_SIMPLEX, 2, (0, 0, 255), 3)
 
+    if time.time() - visual_notification[1] < time_threshold * 2:
+        cv2.putText(frame, visual_notification[0], (frame.shape[1] // 2 + 250, frame.shape[0] // 2 + 250), cv2.FONT_HERSHEY_SIMPLEX, 2, (255, 0, 0), 3)
+
```
Comment (kmsgm) [inline test.py:305]: Is this `cv2.putText(frame, visual_notification[0], (frame.shape[1] // 2 + 250, frame.shape[0] // 2 + 250), cv2.FONT_HERSHEY_SIMPLEX, 2, (255, 0, 0), 3)` different from Print Gesture?

Comment (sangbin-sangbin) [reply test.py:305]: this is for recognized gesture. I will add annotation.

File: test.py (hunk 8 above)
Comment (sangbin-sangbin) [reply test.py:21]: 딱히 오류 날 일이 없어서 그냥 지웠고 while 지운거는 저게 있으면 소리나는동안 다른 코드 진행이 멈춰서 지웠어요

File: test.py (hunk 7 above)
Comment (sangbin-sangbin) [reply test.py:171]: fixed

File: test.py (hunk 9 above)
Comment (sangbin-sangbin) [reply test.py]: fixed

File: test.py (hunk 8 above)
Comment (kmsgm) [reply test.py:21]: Got it
